### PR TITLE
Feature common epistemic overlap

### PR DESCRIPTION
**Pull request by jajapuramshivasai**
## Changes

  -  [ ] Added common_epistemic_overlap.py to implement the feature.
  -  [ ] Added test_common_epistemic_overlap.py with unit tests to validate the functionality.

## Checklist
Before marking your PR ready for review, make sure you checked the following locally. If this is your first PR, you might be notified of some workflow failures after a maintainer has approved the workflow jobs to be run on your PR. 

Additional information is available in the [documentation](https://toqito.readthedocs.io/en/latest/contributing.html#testing).

  -  [x] Use `ruff` for errors related to code style and formatting.
  -  [x] Verify all previous and newly added unit tests pass in `pytest`.
  -  [x] Check the documentation build does not lead to any failures. `Sphinx` build can be checked locally for any failures related to your PR
  -  [x] Use `linkcheck` to check for broken links in the documentation
  -  [x] Use `doctest` to verify the examples in the function docstrings work as expected.

### CODE REVIEW

File: toqito/state_props/common_epistemic_overlap.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    :param state: Input state to check
+    :return: True if state is a vector, False otherwise
+
+    """
+    if state.ndim == 1:
+        return True
+    return state.ndim == 2 and (state.shape[0] == 1 or state.shape[1] == 1)
+
+
+def _vector_to_density_matrix(state: np.ndarray) -> np.ndarray:
```
Comment (purva-thakre) [inline toqito/state_props/common_epistemic_overlap.py:82]: This function already exists in toqito. `vec` and `unvec`

File: toqito/state_props/common_epistemic_overlap.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+def _is_state_vector(state: np.ndarray) -> bool:
+    """Check if input is a state vector (1D or column/row vector).
+
+    :param state: Input state to check
+    :return: True if state is a vector, False otherwise
+
+    """
+    if state.ndim == 1:
+        return True
+    return state.ndim == 2 and (state.shape[0] == 1 or state.shape[1] == 1)
```
Comment (purva-thakre) [inline toqito/state_props/common_epistemic_overlap.py:79]: You can use the `vec` function that already exists in toqito.

File: toqito/state_props/common_epistemic_overlap.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+def _generalized_pauli_Z(d: int) -> np.ndarray:
+    """Generate generalized Pauli Z (clock) operator for dimension d.
+
+    :param d: Dimension of the qudit system
+    :return: d x d generalized Pauli Z operator
+
+    """
+    omega = np.exp(2j * np.pi / d)
+    return np.diag([omega**i for i in range(d)])
```
Comment (purva-thakre) [inline toqito/state_props/common_epistemic_overlap.py:187]: There exists functions in toqito to create generalized versions of Paulis

File: toqito/state_props/common_epistemic_overlap.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    :return: True if n is prime, False otherwise
+
+    """
+    return n > 1 and all(n % i for i in range(2, int(np.sqrt(n)) + 1))
+
+
+def _prime_factors(n: int) -> List[int]:
+    """Compute the prime factorization of n.
+
+    :param n: Number to factorize
```
Comment (purva-thakre) [inline toqito/state_props/common_epistemic_overlap.py:203]: I haven't yet looked at the theory behind this function. But can you briefly explain the need for these functions?

File: toqito/state_props/common_epistemic_overlap.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+from itertools import product
+from typing import List, Union
+
+import numpy as np
+
+
+def common_epistemic_overlap(
+    states: List[np.ndarray], dim: Union[int, List[int], np.ndarray] = None
+) -> float:
+    r"""Compute the epistemic overlap :cite:`Campos_2024_Epistemic`.
```
Comment (purva-thakre) [inline toqito/state_props/common_epistemic_overlap.py:13]: There is something wrong with the citation. It shows up as 

image

But if I follow the arxiv link, the authors on the paper are different than what's in the docs build. 

image

File: toqito/state_props/common_epistemic_overlap.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+    :param state: Input state vector
+    :return: Corresponding density matrix
+
+    """
+    vector = state.reshape(-1)  # Flatten to 1D array
+    return np.outer(vector, vector.conj())
+
+
+def _epistemic_distribution(rho: np.ndarray, vertices: List[np.ndarray]) -> np.ndarray:
```
Comment (vprusso) [inline toqito/state_props/common_epistemic_overlap.py:93]: I don't follow why this needs to be its own function. Also, please use native types.

File: toqito/state_props/common_epistemic_overlap.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        else:
+            if state.shape[0] != state.shape[1]:
+                raise ValueError("Density matrices must be square")
+            density_matrices.append(state)
+    dims = [dm.shape[0] for dm in density_matrices]
+    if len(set(dims)) > 1:
+        raise ValueError("All states must have consistent dimension")
+    vertices = _generate_phase_point_operators(dims[0])
+    distributions = [_epistemic_distribution(dm, vertices) for dm in density_matrices]
+    return np.sum(np.min(distributions, axis=0))
```
Comment (vprusso) [inline toqito/state_props/common_epistemic_overlap.py:67]: Can you explain your logic for how this works? I'm not sure I understand.

File: toqito/state_props/common_epistemic_overlap.py (hunk 3 above)
Comment (vprusso) [inline toqito/state_props/common_epistemic_overlap.py:187]: These functions already exist.

File: toqito/state_props/common_epistemic_overlap.py (hunk 4 above)
Comment (vprusso) [inline toqito/state_props/common_epistemic_overlap.py:203]: I'm not sure why these functions need to be here and also, we would likely never want to define our own functions for these things which would be fairly standard and provided in something like numpy (for instance).

**Review by vprusso — COMMENTED** [review]
Hi @jajapuramshivasai ,

I'm not quite following the updates you've put here. I left some comments, but I'm overall quite a bit confused about your approach and logic.

File: toqito/state_props/common_epistemic_overlap.py (hunk 2 above)
Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py:79]: ill update it

File: toqito/state_props/common_epistemic_overlap.py (hunk 4 above)
Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py:203]: In the chosen epistemic model we define epistemic state given as probability distribution over stabiliser states of the system given by mu(lambda/psi)  in paper , but we can generate these lambda states only for prime dimension system d so to construct  lambda space we have to factories d into primes then we get Lambda states of the composite system as combinations of tensor product of its prime factor lambda space

Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py:203]: I have taken these from implementiation universal wigner function so it is like that,  but ill update the code with predefined functions

File: toqito/state_props/common_epistemic_overlap.py (hunk 7 above)
Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py:67]: basically in this model lambda states are given by vertices and we do projection of our required state given by rho or psi  to get the probability distribution which is out required epistemic state mu(lambda/rho) to find epistemic overlap we take then common area of all probability distributions generated  by function epistemic_distribution for all given states (see definition 2)

File: toqito/state_props/common_epistemic_overlap.py (hunk 6 above)
Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py:93]: I thought it would be more clear and useful to keep them separately because users might want to generate epistemic states for a given single quantum state and there are some other epistemic models in which definition or method of finding epistemic state differs so I wanted to keep that freedom to change . thank you

File: toqito/state_props/common_epistemic_overlap.py (hunk 5 above)
Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py:13]: that must have been an error I have tried to refer to speakers model ,ill update it to refer original paper with correct link

File: toqito/state_props/common_epistemic_overlap.py (hunk 3 above)
Comment (vprusso) [reply toqito/state_props/common_epistemic_overlap.py:187]: Sorry, I should be more specific, they exist in `gen_pauli_x.py` and `gen_pauli_z.py`.

Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py:187]: hi there seems to be a problem with those functions and I they need to be updated accordingly

---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
[<ipython-input-2-7a41950daddf>](https://atucud0jow-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab_20250324-060205_RC00_739881696#) in <cell line: 0>()
      4 from itertools import product
      5 from toqito.matrix_ops.vec import vec
----> 6 from toqito.matrices import gen_pauli_x, gen_pauli_z
      7 
      8 

10 frames
[/usr/local/lib/python3.11/dist-packages/numpy/_core/strings.py](https://atucud0jow-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab_20250324-060205_RC00_739881696#) in <module>
     12 from numpy._core.multiarray import _vec_string
     13 from numpy._core.overrides import set_module
---> 14 from numpy._core.umath import (
     15     isalpha,
     16     isdigit,

ImportError: cannot import name '_center' from 'numpy._core.umath' (/usr/local/lib/python3.11/dist-packages/numpy/_core/umath.py)

---------------------------------------------------------------------------
NOTE: If your import is failing due to a missing package, you can
manually install dependencies using either !pip or !apt.

To view examples of installing some common dependencies, click the
"Open Examples" button below.
---------------------------------------------------------------------------

Comment (purva-thakre) [reply toqito/state_props/common_epistemic_overlap.py]: Do you mind providing your full code for where the error occurs? I just tested the examples in the documentation locally, and they work without any issues. 

```py
from toqito.matrices import gen_pauli_x, gen_pauli_z
matz = gen_pauli_z(3)
matx = gen_pauli_x(4)

print(matx)
print(matz)
```

Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py]: its working on my local machine but when I was testing on google colab and it did not work. please look into it. thank you

Comment (purva-thakre) [reply toqito/state_props/common_epistemic_overlap.py]: I can't replicate it, sorry. Since you used the function in your main function and the tests didn't fail, this might be a local issue for you.

File: toqito/state_props/common_epistemic_overlap.py (hunk 1 above)
Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py]: done

File: toqito/state_props/common_epistemic_overlap.py (hunk 3 above)
Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py]: hi see this https://colab.research.google.com/drive/1wARD7b9dM-Cr65xxGrMXKnqfHoppdkAD?usp=sharing , I think code should be updated to work with newer versions of numpy .

File: toqito/state_props/common_epistemic_overlap.py (hunk 4 above)
Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py]: > I'm not sure why these functions need to be here and also, we would likely never want to define our own functions for these things which would be fairly standard and provided in something like numpy (for instance).

I have replaced those functions with built in functions from sympy , now everything should be sorted.

File: toqito/state_props/common_epistemic_overlap.py (hunk 3 above)
Comment (purva-thakre) [reply toqito/state_props/common_epistemic_overlap.py]: What numpy version are you working with? 

https://github.com/vprusso/toqito/blob/3cc8fe2d5d1ee414947f8dcc4c978133cdd2ab05/pyproject.toml#L37

Comment (vprusso) [reply toqito/state_props/common_epistemic_overlap.py]: > I think code of gen_pauli_x and gen_pauli_z should be updated to work with newer versions of numpy .

Can you explain how these are not compatible with the newer versions of `numpy`?

Comment (vprusso) [reply toqito/state_props/common_epistemic_overlap.py]: Ah, sorry, I see the error you have above. I am also not able to replicate this `_center` issue. I think as @purva-thakre mentions, perhaps this is a local issue for you?

Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py]: I was using 2.0.2 now when I have updated to 2.21 everything is working fine so no issues , thank you

File: toqito/state_props/common_epistemic_overlap.py (hunk 7 above)
Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py]: please see https://arxiv.org/html/2110.12318v3 , thank you for your time

Comment (vprusso) [reply toqito/state_props/common_epistemic_overlap.py]: @jajapuramshivasai I'm not sure it makes sense to obtain negative overlap values. For instance, for a set of randomly generated states, I seem to obtain:

```py
>>> from toqito.rand import random_states
>>> n = 3
>>> d = 3
>>> states = random_states(n, d)
>>> print(common_epistemic_overlap(states))
-0.4530932516083247
```

Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py]: it is given that it is a quasi probability distribution so that maybe the reason for the negative overlap but it could be easily be fixed by taking absolute value of projections , so I would updated the code accordingly

Comment (vprusso) [reply toqito/state_props/common_epistemic_overlap.py]: What about the following scenario though:

```py
from toqito.rand import random_states
from toqito.state_opt import state_exclusion
n = 1
d = 2
states = random_states(n, d)
A, _ = state_exclusion(states)
print(common_epistemic_overlap(states))
print(A)
print(n * (1 - A))
```

The outputs we get here are:

```
0.9999999999999999
0.9999999999999999
1.1102230246251565e-16
```

But according to Equation (6) in [arXiv:2401.17980](https://arxiv.org/pdf/2401.17980), shouldn't the value of `A` be less than or equal to the value of `n * (1 - A)`?

Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py]: but isn't A = (1-state_exclusion/n) ?

Comment (vprusso) [reply toqito/state_props/common_epistemic_overlap.py]: Yes, but:

```py
from toqito.rand import random_states
from toqito.state_opt import state_exclusion
n = 1
d = 3
states = random_states(n, d)
A, _ = state_exclusion(states)
omega = common_epistemic_overlap(states)
eq = n * (1 - A/n)
print(f"{A=}")
print(f"{eq=}")
print(f"{omega=}")
print(omega <= eq)
```

which yields

```
A=0.9999999999999999
eq=1.1102230246251565e-16
omega=np.float64(1.0)
False
```

Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py]: we know omega_q = n*(1-A_q) : (7)  and  A_q  , _ = 1 - state_exclusion(n_states)/n
=> omega_q =  state_exclusion(n_states)
required condition  =>  omega_e < omega_q
that is   common_epistemic_overlap(states) < state_exclusion(states)

Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py]: for single state any overlap should be 1

File: toqito/state_props/common_epistemic_overlap.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        inv2 = pow(2, -1, d)
+        omega = np.exp(2j * np.pi / d)
+        ops = []
+        for q, p in product(range(d), repeat=2):
+            op = (omega ** ((q * p * inv2) % d) *
+                  np.linalg.matrix_power(X, q) @
+                  np.linalg.matrix_power(Z, p) @ D) / d
+            ops.append(op)
+        return ops
+    def generate_phase_point_operators(d):
```
Comment (purva-thakre) [inline toqito/state_props/common_epistemic_overlap.py:100]: why the need to define all these functions?

File: toqito/state_props/common_epistemic_overlap.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+
+def common_epistemic_overlap(states, dim=None) -> float:
+    r"""Compute the epistemic overlap :cite:`Sagnik_2024_Epistemic`.
+
+    For a set of quantum states :math:`\{\rho_i\}`, the epistemic overlap is defined as:
+
+    .. math::
+        \omega_E(\rho_1,\ldots,\rho_n) = \int \min_{\lambda\in\Lambda}
+        (\mu(\lambda|\rho_1), \ldots, \mu(\lambda|\rho_n)) d\lambda
```
[the remark this answers is not in the document]
Comment (purva-thakre) [reply toqito/state_props/common_epistemic_overlap.py:20]: In a few words, describe what epistemic overlap is...

File: toqito/state_props/common_epistemic_overlap.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            v = vec(state)
+            density_matrices.append(np.outer(v, v.conj()))
+        else:
+            if state.shape[0] != state.shape[1]:
+                raise ValueError("Density matrices must be square")
+            density_matrices.append(state)
+
+    dims = [dm.shape[0] for dm in density_matrices]
+    if len(set(dims)) > 1:
+        raise ValueError("All states must have consistent dimension")
```
Comment (purva-thakre) [inline toqito/state_props/common_epistemic_overlap.py:67]: This could be rephrased. What do you mean by consistent dimension?

File: toqito/state_props/tests/test_common_epistemic_overlap.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    ([np.array([[1, 0], [0, 0], [0, 0]]), np.array([[1, 0], [0, 0], [0, 0]])], ValueError),
+
+    #non consistent dimensions
+    ([np.eye(2), np.eye(3)], ValueError),
+
+])
+def test_input_validation(invalid_input, expected_error):
+    """Verify proper error handling for invalid inputs."""
+    with pytest.raises(expected_error):
+        common_epistemic_overlap(invalid_input)
```
Comment (purva-thakre) [inline toqito/state_props/tests/test_common_epistemic_overlap.py:67]: Check if the expected error msg is provided. See below for an example:

https://github.com/vprusso/toqito/blob/416ace6184700be35eed70c3db936057a25985f1/toqito/channel_metrics/tests/test_diamond_distance.py#L25-L41

**Review by purva-thakre — CHANGES_REQUESTED** [review]

File: toqito/state_props/common_epistemic_overlap.py (hunk 10 above)
Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py:64]: I mean it to have equal dimensions that is if the first quantum preparations is given by n qudit all other preparations must  also have same Hilbert space

File: toqito/state_props/common_epistemic_overlap.py (hunk 8 above)
Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py:97]: to find epistemic state of a given quantum state for which we need to find projection on stabiliser states so to generate these projections for any dimensions d we need the functions mentioned for this task

File: toqito/state_props/tests/test_common_epistemic_overlap.py (hunk 11 above)
Comment (jajapuramshivasai) [reply toqito/state_props/tests/test_common_epistemic_overlap.py]: done

File: toqito/state_props/common_epistemic_overlap.py (hunk 8 above)
Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py]: I am going to remove these functions as I would like to change underlying epistemic model such that probabilities are non negative

File: toqito/state_props/common_epistemic_overlap.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jajapuramshivasai) [reply toqito/state_props/common_epistemic_overlap.py:22]: I have added definition and description of each term please check it out , thank you.

File: toqito/state_props/common_epistemic_overlap.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+            distribution[lambda_point] = prob
+            total_prob += prob
+        for lambda_point in lambda_space:
+            distribution[lambda_point] /= total_prob
+        distributions.append(distribution)
+    min_probabilities = {}
+    for lambda_point in lambda_space:
+        min_probabilities[lambda_point] = min(dist.get(lambda_point, 0) for dist in distributions)
+    overlap = sum(min_probabilities.values())
+    return overlap
```
Comment (vprusso) [inline toqito/state_props/common_epistemic_overlap.py:91]: I'm not sure I completey understand what this code is doing. It's a bit difficult to follow and I think it could benefit from more description. It's also quite hard to read as it's all currently bunched together and likely can be written in a way that make it easier to read.

File: toqito/state_props/tests/test_common_epistemic_overlap.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
+"""Unit tests for the common_epistemic_overlap function."""
+
+import numpy as np
+import pytest
+
+from toqito.state_props import common_epistemic_overlap
+
+
+@pytest.mark.parametrize(
+    "states, expected_overlap",
```
Comment (vprusso) [inline toqito/state_props/tests/test_common_epistemic_overlap.py:10]: Are there any cases in which the input to this problem will yield a value other than 0 or 1? I feel that these test cases may not be completely comprehensive or representative of many of the edge cases.

File: toqito/state_props/common_epistemic_overlap.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
+"""Calculate the common epistemic overlap of quantum states."""
+
+from itertools import product
+
+import numpy as np
+
+from toqito.matrix_ops.vec import vec
+
+
+def common_epistemic_overlap(states, dim=None) -> float:
```
Comment (vprusso) [inline toqito/state_props/common_epistemic_overlap.py:10]: This function needs to include type-hints.

Comment (vprusso) [inline toqito/state_props/common_epistemic_overlap.py:10]: Is it necessary to include the dimension as input? Is this not a property of the `states` that are provided to this function?

File: toqito/state_props/common_epistemic_overlap.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    :return: Common epistemic overlap value between 0 and 1
+
+    """
+    density_matrices = []
+    for state in states:
+        if state.ndim == 1 or (state.ndim == 2 and (state.shape[0] == 1 or state.shape[1] == 1)):
+            v = vec(state)
+            density_matrices.append(np.outer(v, v.conj()))
+        else:
+            if state.shape[0] != state.shape[1]:
```
Comment (vprusso) [inline toqito/state_props/common_epistemic_overlap.py:64]: There is already a function to check whether a matrix is square `toqito.matrix_props.is_square`.

File: toqito/state_props/common_epistemic_overlap.py [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+
+    :param states: List of quantum states (vectors or density matrices)
+    :param dim: Optional dimension specification for composite systems
+    :raises ValueError: For invalid inputs or unsupported dimensions
+    :return: Common epistemic overlap value between 0 and 1
+
+    """
+    density_matrices = []
+    for state in states:
+        if state.ndim == 1 or (state.ndim == 2 and (state.shape[0] == 1 or state.shape[1] == 1)):
```
Comment (vprusso) [inline toqito/state_props/common_epistemic_overlap.py:60]: There is already a function in toqito for converting things to density matrices instead of hard-checking like this. Refer to `matrix_ops.calculate_vector_matrix_dimension`.

File: toqito/state_props/tests/test_common_epistemic_overlap.py (hunk 13 above)
Comment (jajapuramshivasai) [reply toqito/state_props/tests/test_common_epistemic_overlap.py:10]: yes I have included a few like 
       (
            [
                np.eye(3) / 3,  # Maximally mixed
                np.outer([1, 0, 0], [1, 0, 0]),  # Pure state |1>
            ],
            0.333333,  # 1/3 overlap from mixed state contribution
        ),
        
if you want I can add more based on inequality epistemic overlap <= 1 - 0.5*tr|rho_1-rho_2|

Comment (vprusso) [reply toqito/state_props/tests/test_common_epistemic_overlap.py:10]: Great, and yes, another example or set of examples would be helpful!

Comment (jajapuramshivasai) [reply toqito/state_props/tests/test_common_epistemic_overlap.py:10]: I added a few and can add more if needed.
